src/lib.rs

```diff
@@ -16,12 +16,41 @@ pub trait Gcd {
     /// assert_eq!(44, 2024u32.gcd(748));
     /// ```
     fn gcd(self, other: Self) -> Self;
+    fn gcd_binary(self, other: Self) -> Self;
+    fn gcd_euclid(self, other: Self) -> Self;
 }
 
 macro_rules! gcd_impl {
     ($($t:ty),*) => ($(
         impl Gcd for $t {
-            fn gcd(self, other: Self) -> Self {
+            fn gcd(self,other: Self) -> Self {
+                self.gcd_binary(other)
+            }
+            fn gcd_binary(self, mut v: Self) -> Self {
+                let mut u = self;
+                if u == 0 {
+                    return v;
+                }
+                if v == 0 {
+                    return u;
+                }
+                let shift = (u | v).trailing_zeros();
+                u >>= shift;
+                v >>= shift;
+                u >>= u.trailing_zeros();
+                loop {
+                    v = v >> (v.trailing_zeros());
+                    if u > v {
+                        core::mem::swap(&mut v, &mut u);
```

Comment:
`mem::swap` is very expensive for two integer types. I used the [xor swap algorithm](https://en.wikipedia.org/wiki/XOR_swap_algorithm) as a test, and the benchmark improved by 2-8%.

Reply:
I actually thought about doing that too, but decided to trust the compiler to be smart, obviously didn't work out ^^

Reply:
At least on my system this seems to make no difference (only 3% improvement only for u8, no measurable improvement for the rest).
This might be because the old code gets optimized to two conditional moves, which on modern processors isn't slow at all.
I would still include the change since on other platforms things might be different and the xor swap shouldn't be slower (even if it uses one instruction more).